snuba_sdk/query_optimizers/or_optimizer.py

```diff
@@ -0,0 +1,68 @@
+from snuba_sdk.conditions import (
+    BooleanCondition,
+    BooleanOp,
+    Condition,
+    Op,
+)
+from snuba_sdk.formula import Formula
+from snuba_sdk.timeseries import Timeseries
+
+
+class OrOptimizer:
+    def optimize(self, query: Formula | Timeseries) -> Formula | Timeseries:
+        if query.filters is None:
+            return query
+
+        optimized = False
+        new_filters = []
+        for cond in query.filters:
+            res = self._optimize_condition(cond)
+            if res is not None:
+                optimized = True
+                new_filters.append(res)
+            else:
+                new_filters.append(cond)
+
+        if not optimized:
+            return query
+        elif isinstance(query, Timeseries):
+            return Timeseries(
+                metric=query.metric,
+                aggregate=query.aggregate,
+                aggregate_params=query.aggregate_params,
+                filters=new_filters,
+                groupby=query.groupby,
+            )
+        else:
+            return Formula(
+                function_name=query.function_name,
+                parameters=query.parameters,
+                aggregate_params=query.aggregate_params,
+                filters=new_filters,
+                groupby=query.groupby,
+            )
+
+    def _optimize_condition(
+        self, cond: BooleanCondition | Condition
+    ) -> BooleanCondition | Condition | None:
+        """
+        Given a condition, returns the optimized version, or None if it can't be optimized
+        """
+        if cond.op != BooleanOp.OR:
+            return None
+        shared_lhs = None
+        rhsides = []
+        for curr in cond.conditions:
+            if (
+                not isinstance(curr, Condition)
+                or curr.op != Op.EQ
+                or (shared_lhs and curr.lhs != shared_lhs)
+            ):
+                # can't be optimized
+                return None
+            if not shared_lhs:
+                shared_lhs = curr.lhs
+            rhsides += [curr.rhs]
+        assert shared_lhs
```

Comment:
Why these two asserts?

Reply:
It was for mypy, I initialized them as None and empty outside of the loop, but they are guaranteed to be set inside the loop.

I can change it to initialize using first element, in order to remove the asserts. Do you think this would be better since it gets rid of the unnecessary asserts?

Reply:
getting rid of unnecessary asserts is best. You don't want a corner case to happen in production and break a bunch of queries